spec/formatting.md

```diff
@@ -370,33 +383,29 @@ The _fallback value_ depends on the contents of the _expression_:
   > `{42 :func}` resolves to the _fallback value_ `|42|` and
   > `{|C:\\| :func}` resolves to the _fallback value_ `|C:\\|`.
 
-- _expression_ with _variable_ _operand_ referring to a local _declaration_ (with or without a _function_):
-  the _value_ to which it resolves (which may already be a _fallback value_)
+- _expression_ with a _variable_ _operand_ referring to a _declaration_:
+  the string representation of the _fallback value_ of the _expression_ of that _declaration_.
 
   > Examples:
   > In a context where `:func` fails to resolve,
-  > the _pattern_'s _expression_ in `.local $var={|val|} {{{$var :func}}}`
-  > resolves to the _fallback value_ `|val|` and the message formats to `{|val|}`.
-  > In a context where `:now` fails to resolve but `:datetime` does not,
-  > the _pattern_'s _expression_ in
+  > the _placeholder_ in `.local $var = {|val| :func} {{{$var}}}`
+  > resolves to the _fallback value_ `|val|`.
+  >
+  > In a context where `:pretty` fails to resolve but `:now` does not,
+  > the _placeholder_ in
   > ```
   > .local $t = {:now format=iso8601}
-  > .local $pretty_t = {$t :datetime}
-  > {{{$pretty_t}}}
+  > {{{$t :pretty}}}
   > ```
-  > (transitively) resolves to the _fallback value_ `:now` and
-  > the message formats to `{:now}`.
+  > resolves to the _fallback value_ `:now`.
```

Comment:
This is a behaviorally distinct change that I hope is not intentional... the current example is attempting to demonstrate _transitive_ fallback such that an a variable defined as output from a defined function that uses as its operand a different variable defined as output from a base function that fails to resolve is the fallback value from that base resolution failure:
```
.local $t = {:now format=iso8601}
.local $pretty_t = {$t :datetime}
{{{$pretty_t}}}
```
formatting as `{:now}` to indicate the broken function reference _through_ the non-broken function expression, whereas the new scenario
```
.local $t = {:now format=iso8601}
{{{$t :pretty}}}
```
in which `:now` resolves but `:pretty` does not fail to capture those details, and AFAIK should format as the non-fallback output from `{:now format=iso8601}`.

At the root of it, I don't see why the fallback value of an expression with a declaration-referencing operand should be the _fallback_ value of that declaration rather than its _resolved_ value. So looking again at
```
.local $t = {:now format=iso8601}
{{{$t :pretty}}}
```
only if _`:now`_ failed to resolve would I expect the result to format as `{:now}`, because otherwise we have a better value for downstream fallbacks—the resolved value of `{:now format=iso8601}`.

Reply:
When reviewing #539, I had not noticed that this behaviour changed. Before that PR, we had this included in the spec: https://github.com/unicode-org/message-format-wg/blob/770cb66a29ee41ae8117cfeba732a266b569c915/spec/formatting.md?plain=1#L284-L288
(note that this was from before introducing input/local, so "local declaration" here meant all declarations)

The language proposed here would indeed revert to the old behaviour, where a failure in a later function would end up using the _fallback value_ of the earliest declaration that its operand is built from, as opposed to the _resolved value_ of its most recent declaration.

The rationalisation for this behaviour is presented in https://github.com/unicode-org/message-format-wg/pull/539#discussion_r1413036522: when any failure happens in the resolution of an _expression_, we should not trust that the resolved value of its operand is by default serialisable. For example, in:
```
.local $user = {$user_id :get-user}
{{Add {$user :first-nam} as friend}}
```
If `:get-user` succeeds but `:first-nam` is a typo that's not caught before we try to format the message, there is a risk that the default serialisation of the `$user` value would reveal information that is not intended to be public. Therefore, the safe formatting result here must be `Add {$user_id} as friend`.

If reverting to this behaviour is not acceptable, then we ought to add a requirement on all _resolved values_ to provide some safe serialization, in case their values are unintentionally included in the formatted result.

Reply:
That's a very reasonable concern that I find compelling. And (assuming agreement from others) the spec should be **very** clear with both text and examples about the consequences—that resolution failures propagate through operands, locking fallback values to the earliest such point in the causal chain regardless of success/failure of downstream functions.

Reply:
I don't get it? I mean, the default serialization of `$user` might reveal inner data of the user object. but the fallback value doesn't serialize that object. It just reveals the _name_ of that object is `$user`. The only time we expose a serialization is when the operand is a literal, in which case the value is already exposed in the message. So there is no danger of `$user`'s content being exposed, only its name.

I guess I'm also wondering why we walk back up the declaration chain to begin with. The example in the spec is:
```
.local $t = {:now format=iso8601}
.local $pretty_t = {$t :datetime}
{{{$pretty_t}}}
```
Shouldn't the fallback value be "at the point of failure"? What's wrong with `{$pretty_t}` coming out? Or `{$t}` if `:datetime` failed? Emiting `:now` doesn't tell you where the failure is and might point at the wrong element in the chain (imagine that `:datetime` were misspelled in the example)

Reply:
Fallbacking is now set to use the fallback value of the last placeholder, rather than looking up a fallback value via its operand.